File: coaxer/resolve_field_type.py
"""Map a coaxer ``Field`` to the Python type used for a Pydantic model.

Used by ``CoaxedPrompt.response_format`` to build a Pydantic model class
from the compiled output schema. Lives as its own module so the
type-mapping rules can evolve (or grow new cases) under tight unit tests
without dragging in the rest of ``prompt.py``.
"""

from __future__ import annotations

from typing import Any, Literal

from coaxer.schema import Field

_PRIMITIVE_TYPES: dict[str, type] = {
    "bool": bool,
    "int": int,
    "float": float,
    "str": str,
}


def resolve_field_type(field: Field) -> Any:
    if field.type == "enum":
        values = field.values or []
        if not values:
            raise ValueError("enum field requires non-empty `values`")
        # Literal subscription with a runtime list is dynamic by nature; ty
        # can't statically verify the values, but pydantic accepts it.
        return Literal[*values]  # ty: ignore[invalid-type-form]
    if field.type in _PRIMITIVE_TYPES:
        return _PRIMITIVE_TYPES[field.type]
    raise ValueError(f"Unsupported field type for response_format: {field.type!r}")

File: packages/javascript/src/coaxed-prompt/resolve-field-type.ts
import { z, type ZodType } from "zod";

export interface MetaField {
  type: string;
  desc?: string | null;
  values?: string[] | null;
}

type Resolver = (field: MetaField) => ZodType;

const RESOLVERS: Record<string, Resolver> = {
  bool: () => z.boolean(),
  int: () => z.number().int(),
  float: () => z.number(),
  str: (f) =>
    f.values && f.values.length > 0 ? z.enum(f.values as [string, ...string[]]) : z.string(),
  enum: (f) => {
    const values = f.values ?? [];
    if (values.length === 0) {
      throw new Error("enum output requires a non-empty `values` list");
    }
    return z.enum(values as [string, ...string[]]);
  },
  bytes: () => {
    throw new Error("Unsupported output type for structured output: bytes");
  },
};

export function resolveFieldType(field: MetaField): ZodType {
  const resolver = RESOLVERS[field.type];
  if (!resolver) throw new Error(`Unknown output type: ${field.type}`);
  return resolver(field);
}
